credential/src/iam_task.rs

```diff
@@ -0,0 +1,111 @@
+use std::env;
+use std::io::Read;
+use std::time::Duration as StdDuration;
+
+use hyper::Client;
+use hyper::header::Connection;
+use serde_json::{self, Value};
+
+use {AwsCredentials, CredentialsError, ProvideAwsCredentials};
+
+const AWS_CREDENTIALS_PROVIDER_IP: &'static str = "169.254.170.2";
+
+/// Provides AWS credentials from a task's IAM role.
+pub struct IamTaskProvider;
+
+impl ProvideAwsCredentials for IamTaskProvider {
+    // TODO: backoff and retry on failure.
+    fn credentials(&self) -> Result<AwsCredentials, CredentialsError> {
+        let aws_container_credentials_relative_uri = match env::var("AWS_CONTAINER_CREDENTIALS_RELATIVE_URI") {
+            Ok(v) => v,
+            Err(_) => return Err(CredentialsError::new("No AWS_CONTAINER_CREDENTIALS_RELATIVE_URI in environment")),
+        };
+
+        let address: String = format!("http://{}{}", AWS_CREDENTIALS_PROVIDER_IP, aws_container_credentials_relative_uri);
+
+        let mut client = Client::new();
+        client.set_read_timeout(Some(StdDuration::from_secs(15)));
+        let mut response = match client.get(&address)
+            .header(Connection::close()).send()
+        {
+            Ok(response) => response,
+            Err(_) => return Err(CredentialsError::new("Couldn't connect to credentials provider")), // add why?
+        };
+
+        let mut body = String::new();
+        if let Err(_) = response.read_to_string(&mut body) {
+            return Err(CredentialsError::new("Didn't get a parsable response body from the credentials provider"));
+        }
+
+        let credentials = try!(parse_credentials_response(&body));
+
+        Ok(credentials)
+    }
+}
+
+fn parse_credentials_response(response: &str) -> Result<AwsCredentials, CredentialsError> {
+    let json_object: Value = match serde_json::from_str(&response) {
+        Ok(v) => v,
+        Err(_) => return Err(CredentialsError::new("Couldn't parse credentials response body.")),
+    };
+
+    fn extract_string_value_from_json(json_object: &Value, key: &str) -> Result<String, CredentialsError> {
+        match json_object.find(key) {
+            Some(v) => Ok(v.as_str().expect(&format!("{} value was not a string", key)).to_owned()),
+            None => Err(CredentialsError::new(&format!("Couldn't find {} in response.", key))),
+        }
+    }
+
+    let access_key_id = try!(extract_string_value_from_json(&json_object, "AccessKeyId"));
+    let secret_access_key = try!(extract_string_value_from_json(&json_object, "SecretAccessKey"));
+    let token = try!(extract_string_value_from_json(&json_object, "Token"));
+    let expiration = try!(extract_string_value_from_json(&json_object, "Expiration"));
+
+    let expiration = try!(expiration.parse());
```

Comment:
Rust sees that we're passing `expiration` to `AwsCredentials::new`, which expects a `DateTime<UTC>`. So it works backwards from there, and assumes that what we want `parse` to return is a `DateTime<UTC>`. There's a specific term for this, IIRC, but it's slipping my mind right now.